### ISSUE TITLE
Unable to deserialize collections

### ISSUE THREAD

**Issue by henryju** [Microsoft/vso-httpclient-java#18]
Any API returning a collection fail to be deserialized.

> Can not deserialize instance of java.util.ArrayList out of START_OBJECT token
>  at [Source: org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream@7586beff; line: 1, column: 1]

This is because [ApiResourceEntityProvider](https://github.com/Microsoft/vso-httpclient-java/blob/0c7a9b8d264231349d17d69992b78ae1bc9b2301/Rest/alm-vss-client/src/main/java/com/microsoft/alm/client/jaxrs/ApiResourceEntityProvider.java#L129) will unwrap collections when the target object is of type java.util.List, but generated code make use of [ArrayList](https://github.com/Microsoft/vso-httpclient-java/blob/0c7a9b8d264231349d17d69992b78ae1bc9b2301/Rest/alm-tfs-client/src/main/generated/com/microsoft/alm/teamfoundation/sourcecontrol/webapi/GitHttpClientBase.java#L16191). And List != ArrayList.

I suggest to fix the generator to replace ArrayList by List (this is a good practice in Java to not expose implementation in APIs).

**Comment by AlexRukhlin** [comment]
Hi,

(Just a side question: Which version of Java do you use?)

As a short workaround we could modify the line in [ApiResourceEntityProvider](https://github.com/Microsoft/vso-httpclient-java/blob/0c7a9b8d264231349d17d69992b78ae1bc9b2301/Rest/alm-vss-client/src/main/java/com/microsoft/alm/client/jaxrs/ApiResourceEntityProvider.java#L129) to support both List and ArrayList.

We'll be glad to receive a Pull Request with your contribution.

Thanks,
Alex

**Comment by henryju** [comment]
I'm using Java 8. I can submit a PR, but I'm still thinking that exposing java.util.List instead of ArrayList would be cleaner. And since this is generated code, I'm not sure I should create a PR to change it. Or is the code of the generator available somewhere?

**Comment by AlexRukhlin** [comment]
I agree with you that exposing java.util.List instead of ArrayList would be cleaner. Just in the current implementation we can not use interfaces in places where deserialization is needed, e.g. in method return types.

You shouldn't change generated classes. You might see two types of sources - ones are generated the others are manually created. ApiResourceEntityProvider is a manually created class.

**Comment by henryju** [comment]
During my tests, I have changed for example the return type of GitHttpClientBase.getPullRequestIterations() to java.util.List and it was working fine. Are we talking about the same thing?

**Comment by AlexRukhlin** [comment]
Yes, we do. You might see that in `GitHttpClientBase.getPullRequestIterations()` the type `ArrayList<GitPullRequestIteration>` is used two times - as the method's return type and as a part of the parameter in the `super.sendRequest` call specifying the deserialization target type, i.e. the **response** type. Sorry, in my previous comment, I used the same term "return type" for both. That was confusing. :-(

Manually, you've changed only the result type. Correct? As I said, in the current implementation of the REST Client Generator, we cannot do that. We're planing its improvement, but but do not have ETA yet. 

That's why I've suggested the short workaround.

**Comment by henryju** [comment]
> Manually, you've changed only the result type. Correct?

No, I changed both. I will prepare a P/R to show you
